lib/io/i2c.cpp

```diff
@@ -0,0 +1,77 @@
+#include "i2c.hpp"
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <sys/ioctl.h>
+
+#if LINUX
+#include <linux/i2c-dev.h>
+#else
+#define I2C_SLAVE 0x0703
+#endif
+
+namespace hyped::io {
+
+I2c::I2c(const uint8_t bus_address)
+    : sensor_address_(0)
+{
+  char path[13];  // up to "/dev/i2c-255"
+  sprintf(path, "/dev/i2c-%d", bus_address);
+  fd_ = open(path, O_RDWR, 0);
+  if (fd_ < 0) { /* log "Could not open i2c device" */ };
+}
+
+I2c::~I2c()
+{
+  close(fd_);
+}
+
+void I2c::setSensorAddress(const uint32_t address)
+{
+  if (fd_ < 0) {
+    /* log "Could not find i2c device"*/;
+    return;
+  }
+
+  sensor_address_ = address;
+  const int ret   = ioctl(fd_, I2C_SLAVE, address);
+  if (ret < 0) {
+    /*log "Could not set sensor address" */;
+    return;
+  }
+}
+
+int I2c::readData(const uint32_t address, uint8_t *data, const size_t len)
```

Comment:
`len` isn't in line with the style guide ... (neither is `fd` so not sure how rigorous you guys are here).

I am also wondering, and this may be going too far, how often do we anticipate that the data is of variable length? Wouldn't it be better to parametrise over the length there? What I'm thinking about is 

```cpp
template<int length>
I2c::ReadResult I2c::readData(const uint32_t address, const std::array<uint8_t, length> data) { /* ... */ }
```

Of course this only works if we always know the size of the message we are sending at compile time. For sensors readings this will be the case!

One needs to be careful not to ruin the callsites (and I don't know how C++ handles this). What we want is something like 

```
i2c.readData(address, message.toArray());
```

Rather than 
```cpp
const auto data = message.toArray();
i2c.readData<data.size()>(address, data);
```

But even in that case the latter is still not any more verbose than your interface!

Reply:
Data length depends on the size of the hardware register (on the sensor/other auxilarry device). Like the temp sensor for example, has 8 bit registers; so to transmit 16 bits of temp data it uses two registers DATA_T_H and DATA_T_L. 

So i don't think we should ever be dealing with variable length data?

Reply:
This is variable at compile time, not at runtime. You would know what sensor your are using by the time you write the code, no?

Reply:
Alright, try the above and see what it looks/feels like?

Reply:
What about this @kshxtij @SnickeyX ?

Reply:
I think we changed it to readByte so the length stuff doesn't really matter